app/src/main/java/com/b_lam/resplash/ui/photo/detail/PhotoDetailActivity.kt

```diff
@@ -310,22 +344,39 @@ class PhotoDetailActivity :
 
     private fun applyWallpaper(uri: Uri) {
         try {
-            startActivity(WallpaperManager.getInstance(this).getCropAndSetWallpaperIntent(uri))
+            startActivity(wallpaperManager.getCropAndSetWallpaperIntent(uri))
         } catch (e: IllegalArgumentException) {
-            var bitmap: Bitmap? = null
-            try {
-                bitmap = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
-                    ImageDecoder.decodeBitmap(ImageDecoder.createSource(contentResolver, uri))
-                } else {
-                    MediaStore.Images.Media.getBitmap(contentResolver, uri)
-                }
-                WallpaperManager.getInstance(applicationContext).setBitmap(bitmap)
-                toast("Wallpaper set successfully")
-            } catch (e: Exception) {
-                error("Error setting wallpaper", e)
+            viewModel.prepareBitmapFromUri(contentResolver, uri)
+            observerWallpaperBitmapResult()
+        }
+    }
+
+    private fun observerWallpaperBitmapResult() {
+        viewModel.wallpaperBitmap.observe(this) { result ->
```

Comment:
You should use an extension function that I wrote called `observeOnce` here so that you don't end up observing the same livedata multiple times if for some reason the user applied the wallpaper multiple times

Reply:
In my perspective, there can be an instance when due to memory constraints at a certain time, the bitmap was not decoded properly. So, if the user again tries to set the wallpaper action will not happen using `observeOnce`.

Reply:
So `observeWallpaperBitmapResult()` gets called each time the user clicks the set wallpaper button. This would result in the activity observing `wallpaperBitmap` multiple times since it `removeObserver()` is never called. The point of `observeOnce` is that it removes the observer automatically once the livedata emits a value. Then when the user clicks the button again,  `observeWallpaperBitmapResult()` gets called and you `observeOnce` `wallpaperBitmap` again.

Reply:
Ohh Yeah, I missed that. I have made the necessary changes.